### PR TITLE
Refactor + improve diagnostics for `&mut T`/`T` mismatch inside Option/Result

### PR DESCRIPTION
**Pull request by clubby789**
Follow up to #114052. This also makes the diagnostics structured + translatable.

r? @WaffleLapkin

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs [hunk 1]
```diff
@@ -1103,65 +1104,46 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                        Applicability::MachineApplicable,
-                    );
+                    let def_path = self.tcx.def_path_str(adt_def.did());
+                    let span = expr.span.shrink_to_hi();
+                    let subdiag = if self.type_is_copy_modulo_regions(self.param_env, ty) {
+                        errors::OptionResultRefMismatch::Copied {
+                            span, def_path
+                        }
+                    } else if let Some(expected_ty_expr) = expected_ty_expr
+                            // FIXME: suggest changes to both expressions to convert both to
```
Comment (clubby789) [inline compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:1123]: Possibly we could have a multipart suggestion to add `.as_ref()` to one side and `.as_deref()` to the other, but that somewhat duplicates an existing diagnostic that adds `.as_deref()`, so not sure of the best solution here

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs [hunk 2]
```diff
@@ -1,8 +1,6 @@
 use super::FnCtxt;
 
-use crate::errors::{
-    AddReturnTypeSuggestion, ExpectedReturnTypeLabel, SuggestBoxing, SuggestConvertViaMethod,
-};
```
Comment (WaffleLapkin) [inline compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:5]: Any particular reason to remove those imports?

**Review by WaffleLapkin — APPROVED** [review]

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:5]: We should be using `errors::` in the code instead of importing them.

Comment (clubby789) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:5]: iirc the main reason being preventing merge conflicts as diagnostic imports tend to get changed a lot

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:5]: Yup

**Comment by WaffleLapkin** [comment]
@bors r+

(I was sure I did this already but apparently not?)

**Pull request merged**